post/webnote-extension/index.md

```diff
@@ -0,0 +1,658 @@
+
+**vue-router** : `vue-router` will load routes like home page, login, signup , edit pages. [Learn more](https://router.vuejs.org/)
+
+## Create Directus plugin
+
+Create new file `plugins/directus.js`
+
+```js
+// plugins/directus.js
+
+import { createDirectus, rest, authentication } from "@directus/sdk";
+
+import Cookies from "js-cookie";
+const directus = createDirectus("https://directus-supabase.onrender.com")
+    .with(
+        authentication("cookie", {
+            autoRefresh: true,
+            credentials: "include",
+            storage: {
+                get() {
+                    if (Cookies.get("directus_auth"))
+                        return JSON.parse(Cookies.get("directus_auth"));
+                    else return null;
+                },
+                set(data) {
+                    Cookies.set("directus_auth", JSON.stringify(data));
+                },
+            },
+        })
+    )
+    .with(rest());
+
+export default directus;
+```
+
+Replace `https://directus-supabase.onrender.com` with your directus url.
+
+Import in `main.js`
+
+```js
+import directus from "./plugins/directus.js";
+
+app.provide("directus", directus);
+```
+
+So chrome extensions can store data in `cookies` or `local storage`. But for security, we will use cookies.
+
+We have used [custom storage](https://docs.directus.io/guides/sdk/authentication.html#configure-custom-storage) feature provided by directus to deal with cookies with different domain.
+
+## Setup Routing
+
+Create new file `plugins/router.js`
+
+```js
+// plugins/router.js
+
+import { createWebHistory, createRouter } from "vue-router";
+import HomeView from "../views/home.vue";
+
+const routes = [
+    { path: "/", name: "home", meta: { public: false }, component: HomeView },
+];
+
+const router = createRouter({
+    history: createWebHistory(),
+    routes,
+});
+
+export default router;
+```
+
+Basic router instance is ready for project. Include in `main.js`
+
+Import in `main.js`
+
+```js
+// src/main.js
+import router from "./plugins/router.js";
+app.use(router);
+```
+
+## Setup CSS
+
+Create new file `src/style.css`
+
+```css
+body {
+    margin: 0;
+    display: flex;
+    place-items: center;
+    min-width: 320px;
+    min-height: 100vh;
+}
+```
+
+Import `css` in `main.js`
+
+```js
+import "./style.css";
+```
+
+## Load extension in browser
+
+Before moving further, lets build the extension and run demo.
+
+-   Add build command in `package.json`
+
+```json
+"build-extension": "vite build && cp manifest.json dist/"
+```
+
+```
+npm run build-extension
+```
+
+-   Open Google Chrome
+-   Go to `chrome://extensions`
+-   Click on `Load Unpacked button` present at top left
+-   select project directory/dist folder
+
+![Load unpacked extension](./images/load-unpacked.png)
+
+![Extension Preview](./images/extension-preview.png)
```

Comment:
````suggestion
## Initialize Extension

Open your terminal and run the following commands to create a new project, install dependencies, and run the project:

```bash
npm create vite@latest directus-webnote-keeper -- --template vue
cd directus-webnote-keeper
npm install
npm install @directus/sdk js-cookie vue-router
npm run dev
```

Add a `manifest.json` file in root directory of the project:

```json
{
    "name": "Directus WebNote Keeper",
    "version": "1.0",
    "manifest_version": 3,
    "author": "Jay Bharadia",
    "description": "Directus WebNote Keeper for capturing urls and store in directus.",
    "icons": {
        "16": "icon.png",
        "32": "icon.png",
        "48": "icon.png",
        "128": "icon.png"
    },
    "action": {
        "default_title": "Directus WebNote Keeper",
        "default_popup": "index.html"
    },
    "permissions": ["activeTab", "storage"]
}
```

You must tell browser about the functionality and permissions required by this extension. You can read more about declaring permissions in the [Google Extensions Docs](https://developer.chrome.com/docs/extensions/develop/concepts/declare-permissions).

- `activeTab`: We need to read the website from current active tab.
- `storage`: Store the authentication token as a cookie.

### Create Directus SDK Plugin

Create new file `plugins/directus.js`:

```js
import { createDirectus, rest, authentication } from "@directus/sdk";

import Cookies from "js-cookie";
const directus = createDirectus("your-project-url")
    .with(
        authentication("cookie", {
            autoRefresh: true,
            credentials: "include",
            storage: {
                get() {
                    if (Cookies.get("directus_auth"))
                        return JSON.parse(Cookies.get("directus_auth"));
                    else return null;
                },
                set(data) {
                    Cookies.set("directus_auth", JSON.stringify(data));
                },
            },
        })
    )
    .with(rest());

export default directus;
```

Replace `your-project-url` with your Directus Project's URL.

Open `main.js` and import the plugin:

```js
import directus from "./plugins/directus.js";
app.provide("directus", directus);
```

### Setup Routing

Create new file `plugins/router.js`

```js
import { createWebHistory, createRouter } from "vue-router";
import HomeView from "../views/home.vue";

const routes = [
    { path: "/", name: "home", meta: { public: false }, component: HomeView },
];

const router = createRouter({
    history: createWebHistory(),
    routes,
});

export default router;
```

Include the router instance in `main.js`:

Import in `main.js`

```js
import directus from "./plugins/directus.js";
import router from "./plugins/router.js"; // [!code ++]
app.provide("directus", directus);
app.use(router); // [!code ++]
```

### Load Extension

Before moving further, build the extension and make sure it runs in your browser. Add build command in `package.json`:

```json
"build-extension": "vite build && cp manifest.json dist/"
```

And then run the build command from your terminal:

```
npm run build-extension
```

Open Google Chrome, go to `chrome://extensions`, click on 'Load Unpacked button' button, and select your project's `dist` folder.
````